streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -0,0 +1,228 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class InternalTopicManager {
+
+    private static final Logger log = LoggerFactory.getLogger(InternalTopicManager.class);
+
+    // TODO: the following ZK dependency should be removed after KIP-4
+    private static final String ZK_TOPIC_PATH = "/brokers/topics";
+    private static final String ZK_BROKER_PATH = "/brokers/ids";
+    private static final String ZK_DELETE_TOPIC_PATH = "/admin/delete_topics";
+
+    private final ZkClient zkClient;
+
+    private class ZKStringSerializer implements ZkSerializer {
+
+        @Override
+        public byte[] serialize(Object data) {
+            try {
+                return ((String) data).getBytes("UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new AssertionError(e);
+            }
+        }
+
+        @Override
+        public Object deserialize(byte[] bytes) {
+            try {
+                if (bytes == null)
+                    return null;
+                else
+                    return new String(bytes, "UTF-8");
+            } catch (UnsupportedEncodingException e) {
+                throw new AssertionError(e);
+            }
+        }
+    }
+
+    public InternalTopicManager(String zkConnect) {
+        zkClient = new ZkClient(zkConnect, 30 * 1000, 30 * 1000, new ZKStringSerializer());
+    }
+
+    public void makeReady(String topic, int numPartitions, int replicationFactor) {
+        Map<Integer, List<Integer>> topicMetadata = getTopicMetadata(topic);
+
+        boolean topicNotReady = true;
+
+        while (topicNotReady) {
+            if (topicMetadata == null) {
+                try {
+                    createTopic(topic, numPartitions, replicationFactor);
+                } catch (ZkNodeExistsException e) {
+                    // ignore and continue
+                }
+            } else {
+                if (topicMetadata.size() > numPartitions) {
+                    // else if topic exists with more #.partitions than needed, delete in order to re-create it
+                    try {
+                        deleteTopic(topic);
+                    } catch (ZkNodeExistsException e) {
+                        // ignore and continue
+                    }
+                } else if (topicMetadata.size() < numPartitions) {
+                    // else if topic exists with less #.partitions than needed, add partitions
+                    try {
+                        addPartitions(topic, numPartitions - topicMetadata.size(), replicationFactor, topicMetadata);
+                    } catch (ZkNoNodeException e) {
+                        // ignore and continue
+                    }
+                }
+
+                topicNotReady = false;
```

Comment:
This is not introduced in this patch, but we should only set `topicNotReady` to false in the `else` condition, i.e. when there is metadata exist, and its numPartitions == topicMetadata.size().